### PR TITLE
feat(gherkin): Added new steps and support for testing activate, get_variation API's and feature-variable api listeners.

### PR DESCRIPTION
**Pull request by yasirfolio3**
### Summary
- Added steps for testing dispatched events of is_feature_enabled api.
- Added steps to process activate and getVariation API's.
- Added support for testing feature-variable api listener's.

### CODE REVIEW

File: tests/integration/main_test.go [hunk 1]
```diff
@@ -64,7 +64,9 @@ func FeatureContext(s *godog.Suite) {
 	s.Step(`^the result should match list "([^"]*)"$`, context.TheResultShouldMatchList)
 	s.Step(`^in the response, "([^"]*)" should be "([^"]*)"$`, context.InTheResponseKeyShouldBeObject)
 	s.Step(`^in the response, "([^"]*)" should match$`, context.InTheResponseShouldMatch)
+	s.Step(`^in the response, "([^"]*)" should have this exactly (\d+) times$`, context.InTheResponseShouldHaveThisExactlyTimes)
```
Comment (msohailhussain) [inline tests/integration/main_test.go]: `ResponseShouldHaveThisExactlyNTimes`

File: tests/integration/optlyplugins/test_composite_service.go [hunk 2]
```diff
@@ -70,17 +70,9 @@ func (c *TestCompositeService) decisionNotificationCallback(notification notific
...
-		decisionInfoDict = notification.DecisionInfo["feature"].(map[string]interface{})
-		source := ""
-		if decisionSource, ok := decisionInfoDict["source"].(decision.Source); ok {
-			source = string(decisionSource)
-		} else {
-			source = decisionInfoDict["source"].(string)
-		}
-		decisionInfoDict["source"] = source
 
+	updateSourceInfo := func(source string) {
```
Comment (msohailhussain) [inline tests/integration/optlyplugins/test_composite_service.go]: Place the condition after switch, no need for this separate method.

File: tests/integration/support/steps.go [hunk 3]
```diff
@@ -170,6 +170,30 @@ func (c *ScenarioCtx) InTheResponseShouldMatch(argumentType string, value *gherk
...
 		break
 	}
+
+	return fmt.Errorf("response for %s not equal", argumentType)
+}
+
+// InTheResponseShouldHaveThisExactlyTimes represents a step in the feature file
+func (c *ScenarioCtx) InTheResponseShouldHaveThisExactlyTimes(argumentType string, count int, value *gherkin.DocString) error {
+	switch argumentType {
+	case "listener_called":
```
Comment (msohailhussain) [inline tests/integration/support/steps.go]: I am not sure, but all these strings should be constant and placed in constant.go
@mikeng13 do u suggest to have those strings in separate file?

File: tests/integration/models/constants.go [hunk 4]
```diff
@@ -36,3 +36,30 @@ const EventProcessorDefaultQueueSize = 1
 
 // EventProcessorDefaultFlushInterval - The default value for event processor flush interval
 const EventProcessorDefaultFlushInterval = 250 * time.Millisecond
+
+// APIType - represents api type
+type APIType string
```
Comment (mikeng13) [inline tests/integration/models/constants.go]: nit: We can call this `SDKAPI` to be more accurate

File: tests/integration/support/steps.go [hunk 5]
```diff
@@ -134,7 +135,8 @@ func (c *ScenarioCtx) TheResultShouldBeBoolean() error {
 
 // TheResultShouldMatchList represents a step in the feature file
 func (c *ScenarioCtx) TheResultShouldMatchList(list string) error {
-	if c.apiResponse.Result == list {
+	expectedList := strings.Split(list, ",")
+	if actualList := c.apiResponse.Result.([]string); compareStringSlice(expectedList, actualList) {
```
Comment (mikeng13) [inline tests/integration/support/steps.go]: nit: Here you should use the `ok` param 

if `actualList, ok := c.apiResponse.Result.([]string); ok { ... }`

File: tests/integration/support/steps.go [hunk 6]
```diff
@@ -170,13 +172,37 @@ func (c *ScenarioCtx) InTheResponseShouldMatch(argumentType string, value *gherk
 	default:
 		break
 	}
+
+	return fmt.Errorf("response for %s not equal", argumentType)
+}
+
+// ResponseShouldHaveThisExactlyNTimes represents a step in the feature file
```
Comment (mikeng13) [inline tests/integration/support/steps.go]: I didn't catch this before but we should be more descriptive about these method docstrings.
 
// ResponseShouldHaveThisExactlyNTimes checks that the response object has the given property exactly N times.

Let's do this for all the other step definitions too please.

File: tests/integration/support/steps.go [hunk 7]
```diff
@@ -192,12 +218,22 @@ func (c *ScenarioCtx) InTheResponseShouldHaveEachOneOfThese(argumentType string,
...
+
 // ThereAreNoDispatchedEvents represents a step in the feature file
 func (c *ScenarioCtx) ThereAreNoDispatchedEvents() error {
-	if len(c.apiResponse.ListenerCalled) == 0 {
+	dispatchedEvents := c.clientWrapper.EventDispatcher.(optlyplugins.EventReceiver).GetEvents()
+	if len(dispatchedEvents) == 0 {
 		return nil
 	}
-	return fmt.Errorf("listenersCalled should be empty")
+	return fmt.Errorf("dispatchedEvents should be empty")
```
Comment (mikeng13) [inline tests/integration/support/steps.go]: nit: Log how many dispatched events we got in addition to saying it should be empty

File: tests/integration/support/utils.go [hunk 8]
```diff
@@ -77,3 +77,29 @@ func parseTemplate(s string, config pkg.ProjectConfig) string {
 	}
 	return parsedString
 }
+
+func compareStringSlice(x, y []string) bool {
```
Comment (mikeng13) [inline tests/integration/support/utils.go]: Did we copy this from somewhere? If so we need proper attribution for this code

Comment (yasirfolio3) [reply tests/integration/support/utils.go]: This was copied from stackoverflow.

Comment (mikeng13) [reply tests/integration/support/utils.go]: Ok I'm that case just include the link to the stack overflow post right above this so we know where it came from

Comment (yasirfolio3) [reply tests/integration/support/utils.go]: done.

Comment (mikeng13) [reply tests/integration/support/utils.go]: @yasirfolio3 I don't see the link, did you forget to commit 😄

Comment (yasirfolio3) [reply tests/integration/support/utils.go]: done, sorry about the confusion, i actually thought you asked me to comment the link here, apologies 👍

**Pull request merged**